### DISCUSSION TITLE
On demand wifi portal

### DISCUSSION THREAD

**Discussion by jayfourjavier** [ESP32-Web/WiFi-Manager_Library#13] [Ideas]
I would suggest to add a builtin function to process on demand wifi configuration (on button press) by using the internal pullup or pulldown resistor of the esp32.

void setupOnDemand(triggerPin, int triggerState,  = default HIGH (1)){
pinMode (triggerPin, triggerState ? INPUT_PULLDOWN: INPUT_PULLUP);
}

void configOnDemand ()
{
if (digitalRead (triggerPin) == triggerState){
if (portal running){
serial print wifi manager is already open
else 
opening wifi manager
}

**Comment by FarazFKhan** [comment]
Hello There,

Thank you for your suggestion. To clarify, are you referring to a way (such as button press) to restart AP Mode after it has timed out (timer run out)?

**Reply by jayfourjavier** [reply]
Yes. And during setup it attempt to connect to any saved ap first, I guess it could save all ap's in the filesystem, so it can try to connect to any of them if there are active APs among them before opening the portal

**Reply by jayfourjavier** [reply]
Do we have updates on this?

**Reply by FarazFKhan** [reply]
Hello, Yes, the the ability of restarting the WiFi Config process has been demoed in the new update. Here is a screenshot of the loop section in the new update. I have added some explanation below as well.

![image](https://github.com/user-attachments/assets/9ea1abb6-1728-48c5-b4ff-8db02e7ad88b)


We can apply conditional logic to restart the WiFi Config process if the chip has entered Idle mode - happens after AP Mode times out. You could further nest an if statement there allowing for a button press trigger for the "checkWiFiConfig()" function to initiate. 

The "checkWiFiConfig()" function runs though the following logic:
>If there are any saved credentials -> attempt connection 
>> -> if successful, start STA Mode
>> -> if fails, start AP Mode

                                                                                            
>If no saved credentials -> start AP Mode

As per your idea of being able to save multiple credentials, where the chips first runs a scan to see if any of the saved credentials match the available SSIDs found through the scan - I have saved this idea to work on later down the road...thanks 👍

**Reply by fbrkt** [reply]
What if you are already connected to a network but you would like want to restart the portal by pressing a button ( or a switch) so to switch to another network ?

**Reply by fbrkt** [reply]
I am suggesting that since The Es32web.local  page worked only on 1 (MacBook Air)  and failed on my 2 other devices ( PC, Android phone)

**Reply by FarazFKhan** [reply]
Firstly, if in STA Mode (station mode), you should be able to just connect to the the wifi router that the ESP32 is connected to and use the WiFi Manager Web App via the default url of "esp32web.local" (unless you have changed this value to a custom one). Image attached below of accessing ESP32 via router - note how it displays the name of the connected router.

![image](https://github.com/user-attachments/assets/b31efebe-c1aa-4ca5-a173-ea4e3d860095)

But if you're keen on restarting AP mode while being on STA mode, you can run some conditional logic in the loop section as follows:

![image](https://github.com/user-attachments/assets/207cbdc0-19eb-4576-aafc-6ebcdbe50ad2)

We have had to remove the existing credentials file, otherwise the "checkWiFiConfig()" function would just use the credentials and reconnect the device to the wifi router, bringing you back to STA Mode which is where you started from. Removing the credentials, and then running the function will allow AP Mode to be established.

I'm not sure why you are only able to access the webpage from one device but not the others. Could you try using a different browser on the devices that wouldn't load the webpage. Another thing you could try, is using the IP address that is printed in the serial monitor after STA mode is established rather than the "esp32web.local" url. Just trying to determine the extent of the issue.

Hope the above helps!

**Reply by jayfourjavier** [reply]
thanks for this, would definitely try this later.

**Reply by fbrkt** [reply]
Thank you for your answer !  

I was thinking more of the case when each time we boot the Esp32, the WIFI manager would automatically connect to the previously saved WIFI (using `checkWiFIConfig()`) but would also let the user the choose to change their network for the next 2 mins or so after boot. 

Which would mean that the Esp32 would be in AP STA mode, then turn back to STA after the timeout

the internet (Kill) switch that I had in mind would reconnect to the WIFI and launch the Portal for 2 mins each time the switch is turned  on, 

and would disconnect to the WIFI and close the portal each time the switch is off . 

the board could also boot with the switch on or off

This is how I was able to implement it so far, idk if that makes sense : 
![image](https://github.com/user-attachments/assets/c3661dd1-40c2-42a4-a269-d3e69f12f108)

It seems to be working well, however there is 2 issues that appears : 

- AP page shows that the board is not connected to a WIFI (even though Wifi.status() says otherwise ) 
 ( Pressing on scan WIFI doesn't change that )
![image](https://github.com/user-attachments/assets/f4f36985-0f2f-4a3c-a360-763240967177)
![image](https://github.com/user-attachments/assets/2d5472af-cf38-45f4-9c36-f5072e75a58e)

-  ~~The board is still in AP+STA mode and not just in STA after the timeout~~  
   the board changes to STA after a few minutes  

and for the Esp32web.local, I was able to launch the page on my PC just using the IP address, however typing the loal url still doesnt work even when I switch browsers . .

**Reply by fbrkt** [reply]
here s the code : 
https://pastebin.com/56JXZwCg

**Reply by FarazFKhan** [reply]
@fbrkt Happy to help you with any issues you're encountering with the ESP32 Web Library. 

Having reviewed your code, I am not sure what the objective of `ConnectWiFI()` & `DisconnectWiFi()` is. If you could briefly explain the overall objective of your code, I could better assist you. 

Looking forward to hearing from you!

**Reply by fbrkt** [reply]
@FarazFKhan I want to be able to activate or deactivate WIFI connection using a switch with my Esp32. 
[Just like this example for WIFI Manager, where the WIFI settup page is called on demand using a button](https://github.com/tzapu/WiFiManager/blob/master/examples/OnDemand/OnDemandWebPortal/OnDemandWebPortal.ino)
I will have different routine depending on if I run my Esp32 on WIFI mode or Offline mode

So what I am trying to do for each case : 
**Switch is "On" : WIFI mode**
- Do not delete the currently saved WIFI credential even if we decide not to stay on the same network
- Try to connect to WIFI using the previously saved credential 
- Start the AP even if the board is now connected to a WiFi network so that the user has the option to switch to another Network for as long as the usual timeout lasts

**Switch is "Off" : Offline mode** 
- Disconnect from WIFI
- immediately stop the AP
- Set the board to `WiFi.mode(WIFI_OFF);` to save power

I hope that clarifies it !

**Reply by FarazFKhan** [reply]
@fbrkt Thank you for explaining. I was hoping to test this out before getting back to you - but doesn't seem like I will get a chance any time soon.

Below is what I think might work.

![image](https://github.com/user-attachments/assets/c6908a99-dc11-494f-93e3-97237fdb819e)



Hope it works!

**Reply by FarazFKhan** [reply]
Just realised that this wouldn't work. Issue is the connection to the WiFi & then starting AP Mode, goes against the grain of the logic of the functions built so far. You'd need to create a custom function for this inside the ESPWeb.cpp file, that you can then evoke in your ConnectWiFi() function. 

I could have a look at creating one as posting here for you to add to your files sometime over the weekend. Just confirm that you are still interested in getting a solution & I'll get back to you next week.

**Reply by fbrkt** [reply]
Thank you for looking into this ! 
yes I would very much be interested in getting a solution. Let me know how I could help, I could give you more details, I also do [have some code that seems to be working](https://pastebin.com/56JXZwCg) . I use it in an external class to handle the wifi switch  and the WIFI_Switch::loop() is something I do include in my main loop

**Comment by Adii-Baadshah** [comment]
Your suggestion is great, especially for making WiFi configuration easier in ESP32 projects. A button press with internal pullup/pulldown resistors is a clean and effective approach.

This can work very well with [10.0.0.1 Piso Wifi Pause](https://pisowifipausetime.ph/) systems. When the user pauses the internet, the on-demand WiFi manager can open to allow quick reconfiguration or reset.

Just a tip — try to keep `setupOnDemand()` and `configOnDemand()` modular and non-blocking, so it doesn't interrupt the system, especially during the pause timer. This makes the 10.0.0.1 Piso Wifi Pause experience smoother and more user-friendly.
